File: ios/HeliosApp/DesignSystem/Motion/HeliosMotion.swift
import SwiftUI

// MARK: - HeliosMotion — spring presets from design-tokens.json

enum HeliosMotion {
    enum Preset {
        case gentle
        case `default`
        case snappy
        case bouncy
    }

    static func spring(_ preset: Preset = .default, duration: TimeInterval? = nil) -> Animation {
        switch preset {
        case .gentle:
            return .spring(response: responseFor(stiffness: 120, damping: 24),
                           dampingFraction: dampingFractionFor(damping: 24, mass: 1, stiffness: 120),
                           blendDuration: duration ?? 0.5)
        case .default:
            return .spring(response: responseFor(stiffness: 200, damping: 26),
                           dampingFraction: dampingFractionFor(damping: 26, mass: 1, stiffness: 200),
                           blendDuration: duration ?? 0.35)
        case .snappy:
            return .spring(response: responseFor(stiffness: 380, damping: 32),
                           dampingFraction: dampingFractionFor(damping: 32, mass: 1, stiffness: 380),
                           blendDuration: duration ?? 0.25)
        case .bouncy:
            return .spring(response: responseFor(stiffness: 280, damping: 14),
                           dampingFraction: dampingFractionFor(damping: 14, mass: 1.2, stiffness: 280),
                           blendDuration: duration ?? 0.6)
        }
    }

    enum Duration {
        static let instant: TimeInterval = 0
        static let fast: TimeInterval = 0.15
        static let normal: TimeInterval = 0.25
        static let slow: TimeInterval = 0.4
        static let deliberate: TimeInterval = 0.6
        static let draw: TimeInterval = 0.6
    }

    enum Stagger {
        static let tight: TimeInterval = 0.03
        static let normal: TimeInterval = 0.05
        static let loose: TimeInterval = 0.08
    }

    // MARK: - Spring math helpers

    private static func responseFor(stiffness: Double, damping: Double) -> Double {
        let mass: Double = 1.0
        let naturalFreq = sqrt(stiffness / mass)
        let dampingRatio = damping / (2 * sqrt(mass * stiffness))
        if dampingRatio >= 1 {
            return 0.3
        }
        return (2 * .pi) / (naturalFreq * sqrt(1 - dampingRatio * dampingRatio))
    }

    private static func dampingFractionFor(damping: Double, mass: Double, stiffness: Double) -> CGFloat {
        let ratio = damping / (2 * sqrt(mass * stiffness))
        return CGFloat(min(ratio, 0.999))
    }
}

// MARK: - Choreographed entrance modifier

struct ChoreographedEntrance: ViewModifier {
    let delay: TimeInterval
    let translateY: CGFloat
    @State private var isVisible = false

    func body(content: Content) -> some View {
        content
            .opacity(isVisible ? 1 : 0)
            .offset(y: isVisible ? 0 : translateY)
            .animation(HeliosMotion.spring(.gentle).delay(delay), value: isVisible)
            .onAppear {
                isVisible = true
            }
    }
}

struct StaggeredEntrance: ViewModifier {
    let index: Int
    let baseDelay: TimeInterval
    let stagger: TimeInterval
    @State private var isVisible = false

    func body(content: Content) -> some View {
        let delay = baseDelay + Double(index) * stagger
        return content
            .opacity(isVisible ? 1 : 0)
            .offset(y: isVisible ? 0 : 12)
            .animation(HeliosMotion.spring(.default).delay(delay), value: isVisible)
            .onAppear {
                DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
                    isVisible = true
                }
            }
    }
}

extension View {
    func choreographedEntrance(delay: TimeInterval = 0, translateY: CGFloat = 12) -> some View {
        modifier(ChoreographedEntrance(delay: delay, translateY: translateY))
    }

    func staggeredEntrance(index: Int, baseDelay: TimeInterval = 0, stagger: TimeInterval = HeliosMotion.Stagger.normal) -> some View {
        modifier(StaggeredEntrance(index: index, baseDelay: baseDelay, stagger: stagger))
    }
}

File: android/app/src/main/java/com/helios/core/designsystem/motion/HeliosMotion.kt
package com.helios.core.designsystem.motion

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.Easing
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.SpringSpec
import androidx.compose.animation.core.spring

/**
 * Motion tokens from `design/tokens.json` (`motion`) and
 * `shared-spec/motion-language.md`.
 *
 * Springs are the default for every transition. A cubic curve is only correct for an
 * opacity cross-fade, an indeterminate loader, or the skeleton shimmer, so the curves
 * below are named for those three jobs only.
 */
object HeliosMotion {

    /** Gentle: dampingRatio=0.85, stiffness=120. Hero cards, page entrance, SoC trim. */
    val gentle: SpringSpec<Float> = spring<Float>(dampingRatio = 0.85f, stiffness = 120f)

    /** Default: dampingRatio=0.8, stiffness=200. Standard transitions. */
    val defaultSpring: SpringSpec<Float> = spring<Float>(dampingRatio = 0.8f, stiffness = 200f)

    /** Snappy: dampingRatio=0.7, stiffness=380. Toggles, tickers, nav. */
    val snappy: SpringSpec<Float> = spring<Float>(dampingRatio = 0.7f, stiffness = 380f)

    /** Bouncy: dampingRatio=0.5, stiffness=280. Success moments and empty-state entrance. */
    val bouncy: SpringSpec<Float> = spring<Float>(dampingRatio = 0.5f, stiffness = 280f)

    /** Named spring lookup, for tables and the gallery. */
    fun spring(preset: Preset): SpringSpec<Float> = when (preset) {
        Preset.GENTLE -> gentle
        Preset.DEFAULT -> defaultSpring
        Preset.SNAPPY -> snappy
        Preset.BOUNCY -> bouncy
    }

    enum class Preset { GENTLE, DEFAULT, SNAPPY, BOUNCY }

    /** Duration tokens in milliseconds, the unit Compose needs. */
    object DurationMs {
        const val instant = 0
        const val fast = 150
        const val normal = 250
        const val slow = 400
        const val deliberate = 600
        const val draw = 600
        const val ringTrim = 800
        const val dischargePulse = 1000
        const val chargeGlowPulse = 1600
        const val nodePulse = 2400
        const val shimmer = 2500
        const val screenEnter = 350
        const val reduceMotionCrossfade = 200
    }

    /** Duration tokens in seconds, kept for callers that already use them. */
    object Duration {
        const val instant = 0.0f
        const val fast = 0.15f
        const val normal = 0.25f
        const val slow = 0.4f
        const val deliberate = 0.6f
        const val draw = 0.6f
    }

    /** Stagger delays in milliseconds (entrance choreography). */
    object StaggerMs {
        const val tight = 30
        const val normal = 50
        const val loose = 80
    }

    /** Stagger delays in seconds, kept for callers that already use them. */
    object Stagger {
        const val tight = 0.03f
        const val normal = 0.05f
        const val loose = 0.08f
    }

    /** Screen entrance order from motion-language section 3. */
    object Entrance {
        const val heroDelayMs = 0
        const val statsStaggerMs = 40
        const val sectionHeaderDelayMs = 200
        const val chartDelayMs = 300
        const val chromeDelayMs = 400
        const val chartDrawMs = 600
    }

    /**
     * The only cubic curves in the app. [CURVES.linear] belongs to indeterminate
     * loaders and the shimmer; the other three to opacity cross-fades.
     */
    object Curves {
        val linear: Easing = LinearEasing
        val standard: Easing = CubicBezierEasing(0.2f, 0f, 0f, 1f)
        val decelerate: Easing = CubicBezierEasing(0f, 0f, 0f, 1f)
        val accelerate: Easing = CubicBezierEasing(0.3f, 0f, 1f, 1f)
    }

    /**
     * Reduce Motion collapses every spring to a short opacity cross-fade
     * (motion-language section 11). Components read this flag rather than re-deciding.
     */
    object ReduceMotion {
        const val crossfadeMs = DurationMs.reduceMotionCrossfade
        const val stopsEnergyFlowDots = true
        const val snapsTickers = true
        const val disablesParallax = true
    }
}
